backend/src/apiserver/metadata/metadata_store.go

```diff
@@ -0,0 +1,119 @@
+package metadata
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	argoWorkflow "github.com/argoproj/argo/pkg/apis/workflow/v1alpha1"
+	"github.com/golang/protobuf/jsonpb"
+	"github.com/golang/protobuf/proto"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+
+	"ml_metadata/metadata_store/mlmetadata"
+	mlpb "ml_metadata/proto/metadata_store_go_proto"
+)
+
+// Store ...
+type Store struct {
+	mlmdStore *mlmetadata.Store
+}
+
+// NewStore ...
+func NewStore(mlmdStore *mlmetadata.Store) *Store {
+	return &Store{mlmdStore: mlmdStore}
+}
+
+// RecordOutputArtifacts ...
+func (s *Store) RecordOutputArtifacts(runID string, storedManifest, currentManifest string) error {
+	storedWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(storedManifest), storedWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	currentWorkflow := &argoWorkflow.Workflow{}
+	if err := json.Unmarshal([]byte(currentManifest), currentWorkflow); err != nil {
+		return util.NewInternalServerError(err, "unmarshaling workflow failed")
+	}
+
+	completed := make(map[string]bool)
+	for _, n := range storedWorkflow.Status.Nodes {
+		if n.Completed() {
+			completed[n.ID] = true
+		}
+	}
+
+	for _, n := range currentWorkflow.Status.Nodes {
+		if n.Completed() && !completed[n.ID] {
+			// Newly completed node. Record output ml-metadata artifacts.
+			if n.Outputs != nil {
+				for _, output := range n.Outputs.Parameters {
+					if !strings.HasPrefix(output.ValueFrom.Path, "/output/ml_metadata/") {
+						continue
+					}
+
+					artifacts, err := parseTFXMetadata(*output.Value)
+					if err != nil {
+						return util.NewInternalServerError(err, "metadata parsing failure")
+					}
+
+					for _, a := range artifacts {
+						id, err := s.mlmdStore.PutArtifactType(
+							a.ArtifactType, &mlmetadata.PutTypeOptions{AllFieldsMustMatch: true})
+						if err != nil {
+							return util.NewInternalServerError(err, "failed to register artifact type")
+						}
+						a.Artifact.TypeId = proto.Int64(int64(id))
+						_, err = s.mlmdStore.PutArtifacts([]*mlpb.Artifact{a.Artifact})
+						if err != nil {
+							return util.NewInternalServerError(err, "failed to record artifact")
+						}
+
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+type artifactStruct struct {
```

Comment:
do we picturing this to be the sole artifact structure or tfx specific? If latter, i would recommend reflect that in the type name